pkg/csi/service/vanilla/controller.go

```diff
@@ -1324,120 +1324,192 @@ func (c *controller) createBlockVolumeWithPlacementEngineForMultiVC(ctx context.
+						scParams.StoragePolicyName, vcHost)
+				}
+
+				// Get shared accessible datastores for topology segments associated with the vcHost.
+				sharedDatastores, err = placementengine.GetSharedDatastores(ctx,
+					placementengine.VanillaSharedDatastoresParams{
+						Vcenter:              vcenter,
+						TopologySegmentsList: topologySegmentsList,
+						StoragePolicyID:      storagePolicyID,
+					})
+				if err != nil {
+					return nil, csifault.CSIInternalFault, logger.LogNewErrorCodef(log, codes.Internal,
+						"failed to get shared datastores for topology segments %+v in vCenter %q. Error: %+v",
+						topologySegmentsList, vcHost, err)
+				}
+				if len(sharedDatastores) == 0 {
+					errMsg := fmt.Sprintf("No compatible datastores found for accessibility requirements %+v "+
+						"pertaining to vCenter %q", topologySegmentsList, vcHost)
+					log.Warn(errMsg)
+					combinedErrMssgs = append(combinedErrMssgs, errMsg)
+					continue
+				}
+				// Filter datastores based on user access.
+				sharedDatastores, err = c.filterDatastores(ctx, sharedDatastores, vcHost)
+				if err != nil {
+					if err == errAllDSFilteredOut {
+						errMsg := fmt.Sprintf("authorization service filtered out all the compatible "+
+							"datastores found for accessibility requirements %+v associated with vCenter %q",
+							topologySegmentsList, vcHost)
+						log.Warn(errMsg)
+						combinedErrMssgs = append(combinedErrMssgs, errMsg)
+						continue
+					}
+					return nil, csifault.CSIInternalFault, logger.LogNewErrorCodef(log, codes.Internal,
+						"failed to filter datastores based on authorisation check in vCenter %q. Error: %+v",
+						vcHost, err)
+				}
+				volumeMgr, err = common.GetVolumeManagerFromVCHost(ctx, c.managers, vcHost)
+				if err != nil {
+					return nil, csifault.CSIInternalFault, logger.LogNewErrorCode(log, codes.Internal, err.Error())
+				}
+				// Call CreateVolume.
+				// TODO: Few errors encountered  in CreateBlockVolumeUtilForMultiVC can be
+				// retried instead of moving unto next VC. Need to throw a custom error for such scenarios.
+				volumeInfo, faultType, err = common.CreateBlockVolumeUtilForMultiVC(ctx,
+					common.VanillaCreateBlockVolParamsForMultiVC{
+						Vcenter:              vcenter,
+						VolumeManager:        volumeMgr,
+						CNSConfig:            c.managers.CnsConfig,
+						StoragePolicyID:      storagePolicyID,
+						Spec:                 &createVolumeSpec,
+						SharedDatastores:     sharedDatastores,
+						SnapshotDatastoreURL: snapshotDatastoreURL,
+						ClusterFlavor:        cnstypes.CnsClusterFlavorVanilla,
+					})
+				if err != nil {
+					log.Error(err)
+					combinedErrMssgs = append(combinedErrMssgs, err.Error())
+					continue
+				}
+				log.Infof("volume %q created in vCenter %q. Proceeding to calculate "+
+					"accessible topology for the volume", volumeInfo.VolumeID.Id, vcHost)
+				break
+			}
+		} else {
```

Comment:
can we have common function for below functionality in this block and re-use it for topology and non-topology section (if block above and else block here) with following steps:
input : vchost, topologySegList, scParams, snapshotDatastoreURL, and others needed...
1. fetch vc & volume manager
2. check for storage policy
3. based on null/non-null topologySegList input, call placementengine.GetSharedDatastores or c.nodeMgr.GetSharedDatastoresInK8SCluster
4. filter datastores based on user access
5. call common.CreateBlockVolumeUtilForMultiVC

Reply:
Can we do this code duplication later? 

We have diff logic for multi-vCenter, in some cases, we are handling errors and attempting to go on another vCenter to provision a volume. Validation is also diff for multi-vCenter vs single VC non-topology. We have preferential datastore selection for topology-aware setup vs no preferential datastore for non-topology setup.  We can not call `common.CreateBlockVolumeUtilForMultiVC` in common for both topology and non-topology, as if we have an error in the case when we have multiple topology segments, we want to go to the next segment to provision volume.